### ISSUE TITLE
DiskReadViolation: PersistedInstallation access disk at constructor

### ISSUE THREAD

**Issue by carlonzo** [firebase/firebase-android-sdk#3313]
### [READ] Step 1: Are you in the right place?
Yes this is a bug in the sdk.

### [REQUIRED] Step 2: Describe your environment

Firebase Installation 17.0.0
'com.google.firebase:firebase-installations:17.0.0'

### [REQUIRED] Step 3: Describe the problem

```
D/StrictMode: StrictMode policy violation; ~duration=2121 ms: android.os.strictmode.DiskReadViolation
        at android.os.StrictMode$AndroidBlockGuardPolicy.onReadFromDisk(StrictMode.java:1596)
        at libcore.io.BlockGuardOs.access(BlockGuardOs.java:71)
        at libcore.io.ForwardingOs.access(ForwardingOs.java:72)
        at android.app.ActivityThread$AndroidOs.access(ActivityThread.java:7858)
        at java.io.UnixFileSystem.checkAccess(UnixFileSystem.java:281)
        at java.io.File.exists(File.java:820)
        at android.app.ContextImpl.ensurePrivateDirExists(ContextImpl.java:692)
        at android.app.ContextImpl.ensurePrivateDirExists(ContextImpl.java:683)
        at android.app.ContextImpl.getFilesDir(ContextImpl.java:728)
        at android.content.ContextWrapper.getFilesDir(ContextWrapper.java:253)
        at com.google.firebase.installations.local.PersistedInstallation.<init>(PersistedInstallation.java:84)
        at com.google.firebase.installations.FirebaseInstallations.<init>(FirebaseInstallations.java:140)
        at com.google.firebase.installations.FirebaseInstallationsRegistrar.lambda$getComponents$0(FirebaseInstallationsRegistrar.java:44)
        at com.google.firebase.installations.FirebaseInstallationsRegistrar$$ExternalSyntheticLambda0.create(Unknown Source:0)
        at com.google.firebase.components.ComponentRuntime.lambda$discoverComponents$0$com-google-firebase-components-ComponentRuntime(ComponentRuntime.java:132)
        at com.google.firebase.components.ComponentRuntime$$ExternalSyntheticLambda1.get(Unknown Source:4)
        at com.google.firebase.components.Lazy.get(Lazy.java:53)
        at com.google.firebase.components.AbstractComponentContainer.get(AbstractComponentContainer.java:27)
        at com.google.firebase.components.ComponentRuntime.get(ComponentRuntime.java:45)
        at com.google.firebase.components.RestrictedComponentContainer.get(RestrictedComponentContainer.java:89)
        at com.google.firebase.iid.Registrar.lambda$getComponents$0$Registrar(com.google.firebase:firebase-iid@@21.1.0:5)
        at com.google.firebase.iid.Registrar$$Lambda$0.create(Unknown Source:0)
        at com.google.firebase.components.ComponentRuntime.lambda$discoverComponents$0$com-google-firebase-components-ComponentRuntime(ComponentRuntime.java:132)
        at com.google.firebase.components.ComponentRuntime$$ExternalSyntheticLambda1.get(Unknown Source:4)
        at com.google.firebase.components.Lazy.get(Lazy.java:53)
        at com.google.firebase.components.AbstractComponentContainer.get(AbstractComponentContainer.java:27)
        at com.google.firebase.components.ComponentRuntime.get(ComponentRuntime.java:45)
        at com.google.firebase.components.RestrictedComponentContainer.get(RestrictedComponentContainer.java:89)
        at com.google.firebase.iid.Registrar.lambda$getComponents$1$Registrar(com.google.firebase:firebase-iid@@21.1.0:1)
        at com.google.firebase.iid.Registrar$$Lambda$1.create(Unknown Source:0)
        at com.google.firebase.components.ComponentRuntime.lambda$discoverComponents$0$com-google-firebase-components-ComponentRuntime(ComponentRuntime.java:132)
        at com.google.firebase.components.ComponentRuntime$$ExternalSyntheticLambda1.get(Unknown Source:4)
        at com.google.firebase.components.Lazy.get(Lazy.java:53)
        at com.google.firebase.components.AbstractComponentContainer.get(AbstractComponentContainer.java:27)
        at com.google.firebase.components.ComponentRuntime.get(ComponentRuntime.java:45)
        at com.google.firebase.components.RestrictedComponentContainer.get(RestrictedComponentContainer.java:89)
        at com.google.firebase.messaging.FirebaseMessagingRegistrar.lambda$getComponents$0(com.google.firebase:firebase-messaging@@23.0.0:3)
        at com.google.firebase.messaging.FirebaseMessagingRegistrar$$ExternalSyntheticLambda0.create(Unknown Source:0)
        at com.google.firebase.components.ComponentRuntime.lambda$discoverComponents$0$com-google-firebase-components-ComponentRuntime(ComponentRuntime.java:132)
        at com.google.firebase.components.ComponentRuntime$$ExternalSyntheticLambda1.get(Unknown Source:4)
        at com.google.firebase.components.Lazy.get(Lazy.java:53)
```

Reading the source code of the constructor of [com.google.firebase.installations.FirebaseInstallations](https://github.com/firebase/firebase-android-sdk/blob/master/firebase-installations/src/main/java/com/google/firebase/installations/local/PersistedInstallation.java) the `dataFile` is created reading the disk. As the construction of this class happens on main thread, this leads to a `DiskReadViolation`. 

In particular, `context.getFilesDir()` ([source code here](https://cs.android.com/android/platform/superproject/+/master:frameworks/base/core/java/android/app/ContextImpl.java;drc=master;bpv=1;bpt=1;l=774?gsn=getFilesDir&gs=kythe%3A%2F%2Fandroid.googlesource.com%2Fplatform%2Fsuperproject%3Flang%3Djava%3Fpath%3Dandroid.app.ContextImpl%23ed04d6eb6ba8a8771ea0dca4b0ac59f920662035b4d8adba2d3a6a47fed3747e) calls `ensurePrivateDirExists` which then access the disk. Probably the initialization of `dataFile` can happen lazily when the file is access later safely on a separate thread

**Comment by carlonzo** [comment]
Hi @argzdev, sure. you probably cannot repro with only those steps because the initialization in Firebase is done using a ContenProvider, before the StrictMode is applied.

To do that, you should disable the initialization from ContentProvider ( [as described here](https://firebase.googleblog.com/2017/03/take-control-of-your-firebase-init-on.html)) and then manually call the Firebase initialization code after the StrictMode is applied `FirebaseApp.initializeApp`

here is a small repro project you can try. hope it helps
[ReproFirebaseInstall.zip](https://github.com/firebase/firebase-android-sdk/files/7870356/ReproFirebaseInstall.zip)

**Comment by argzdev** [comment]
Thanks for the repro, @carlonzo. I was able to encounter the issue, with that said, I'll notify an engineer about the PR and see if it's clear for merge for if any update is needed.

**Comment by carlonzo** [comment]
Closed by #3314
